src/dos/cdrom_ioctl_linux.cpp

```diff
@@ -113,4 +320,159 @@ bool CDROM_Interface_Ioctl::ReadSectorsHost([[maybe_unused]] void *buffer,
 	return false; /*TODO*/
 }
 
+void CDROM_Interface_Ioctl::InitAudio(int device_number)
+{
+	if (mixer_channel) {
+		return;
+	}
+
+	std::string name = std::string(MixerChannelNamePrefix) +
+	                   std::to_string(device_number);
+	audio_buffer.resize(AudioBufferBytes / 2);
+
+	auto callback = std::bind(&CDROM_Interface_Ioctl::CDAudioCallback,
+	                          this,
+	                          std::placeholders::_1);
+	mixer_channel = MIXER_AddChannel(callback,
+	                                 REDBOOK_PCM_FRAMES_PER_SECOND,
+	                                 name.c_str(),
+	                                 {ChannelFeature::Stereo,
+	                                  ChannelFeature::DigitalAudio});
+}
+
+void CDROM_Interface_Ioctl::CDAudioCallback(uint16_t requested_frames)
+{
+	if (audio_position == 0) {
+		if (sectors_remaining < 1) {
+			StopAudio();
+			return;
+		}
+		cdrom_read_audio cd = {};
+		cd.addr.lba         = current_sector;
+		cd.addr_format      = CDROM_LBA;
+		cd.buf = reinterpret_cast<uint8_t*>(audio_buffer.data());
+
+		// Maximum this ioctl can read in a single call is 1 second (75
+		// redbook frames)
+		cd.nframes = std::min(sectors_remaining, CD_FRAMES);
+		constexpr int PcmFramesPerRedbookFrame = BYTES_PER_RAW_REDBOOK_FRAME /
+		                                         BYTES_PER_REDBOOK_PCM_FRAME;
+		frames_in_buffer = cd.nframes * PcmFramesPerRedbookFrame;
+		current_sector += cd.nframes;
+		sectors_remaining -= cd.nframes;
+		if (ioctl(cdrom_fd, CDROMREADAUDIO, &cd) != 0) {
+			// Write out silence if something goes wrong
+			std::fill(audio_buffer.begin(), audio_buffer.end(), 0);
+
+#ifdef DEBUG_IOCTL
+			LOG_WARNING("CDROM_IOCTL: CDAudioCallback: CDROMREADAUDIO ioctl failed");
+#endif
+		}
+	}
+
+	const size_t frames_available = frames_in_buffer - audio_position;
+	const size_t frames_to_write = std::min(static_cast<size_t>(requested_frames),
+	                                        frames_available);
+
+	// audio_position must be doubled since a frame is two int16_t's for stereo sound
+#ifdef WORDS_BIGENDIAN
+	mixer_channel->AddSamples_s16_nonnative(frames_to_write,
+	                                        audio_buffer.data() +
+	                                                (audio_position * 2));
+#else
+	mixer_channel->AddSamples_s16(frames_to_write,
+	                              audio_buffer.data() + (audio_position * 2));
+#endif
+
+	audio_position += frames_to_write;
+	if (audio_position >= frames_in_buffer) {
```

Comment:
Yeah so this gets confusing... `audio_position` seems to be a "frame position"... but it's not called that.

Reply:
renamed to `frame_position`